### PR TITLE
Scenario Caching

### LINKED ISSUE #1595: Scenario Studio: Metadata Caching
Issue:
Scenario building takes a long time to do.

Request:
SMARTS should cache scenario builds and only rebuild components as necessary.

Consideration:
Build dependencies may be a direction we may want to go.

e.g. map_build->scenario_build->missions_build

AC:
- [ ] Build dependency graph is documented
- [ ] File modification date taken into account
- [ ] SMARTS version taken into account

### PR DESCRIPTION
**Pull request by saulfield**
Closes #1595.

## Summary of Changes

This PR makes a number of substantial changes to scenario builds:

- Scenario build artifacts are now cached and built incrementally, meaning that subsequent builds (without the `clean` option) will only build the artifacts that depend on the changed DSL objects
- If the map is a local file in the scenario directory (e.g. a `map.net.xml` or `.xodr` file), changes to the map file itself will also trigger a rebuild of artifacts that depend on the map
- All build artifacts are now in a local `build/` directory in each scenario's directory
- A `build.db` SQLite database file is now created during builds to keep track of dependencies
- The `allow_offset_map` option has been removed. This must now be set by setting `shift_to_origin=True` in a `MapSpec` object in the scenario.py file if this option is needed. 
- All scenarios must have a `scenario.py`, and must call `gen_scenario()`, rather than the individual `gen_` functions. Those functions have been made private, and a deprecation warning will be raised if called from outside of `gen_scenario()`.
- Empty meshes are now skipped when generating the map GLB file to get rid of noisy trimesh warnings
- Map GLB file is now only built once (unless the map changes)

## Example

Here is the output from a terminal session that demonstrates what a typical iteration workflow looks like now:

```sh
# First, clean and build the scenario
$ time scl scenario build --clean scenarios/sumo/minicity/
build-scenario scenarios/sumo/minicity/
INFO:/home/saul/code/SMARTS/smarts/sstudio/genscenario.py:"map_glb" took: 32551.769018ms
INFO:/home/saul/code/SMARTS/smarts/sstudio/genscenario.py:"traffic" took: 8146.712780ms
INFO:/home/saul/code/SMARTS/smarts/sstudio/genscenario.py:"social_agent_missions" took: 5.387068ms
INFO:/home/saul/code/SMARTS/smarts/sstudio/genscenario.py:"bubbles" took: 3.000498ms

real	0m42.086s
user	0m42.094s
sys	0m2.561s

# Then, make a change to the traffic specification in scenario.py, and rebuild (without --clean)
$ time scl scenario build scenarios/sumo/minicity/
build-scenario scenarios/sumo/minicity/
INFO:/home/saul/code/SMARTS/smarts/sstudio/genscenario.py:"traffic" took: 7710.859299ms

real	0m9.068s
user	0m9.289s
sys	0m2.256s
```

### CODE REVIEW

**Comment by qianyi-sun** [comment]
Hi Saul, just minor, have you tested the this in the replay scenarios (e.g., NGSIM)? I went though your changes/descriptions and the current `scenario.py` files in the NGSIM folder, looks ok but just wanted to double check.

Also regarding the following point:
"The `allow_offset_map` option has been removed. This must now be set by setting `shift_to_origin=True` in a `MapSpec` object in the scenario.py file if this option is needed."
Should we check all the replay scenarios to see if we should add the `shift_to_origin=True` if we havent done yet? I can check them once this PR get merged. 

Thanks!

File: smarts/sstudio/__init__.py
Comment (Gamenot) [reply smarts/sstudio/__init__.py:34]: the alternative is to do as you did here and make the private methods public in `genscenario`.

**Comment by Gamenot** [comment]
@Gamenot  I need to do a second review specifically on the build changes.

File: smarts/core/utils/file.py
Comment (saulfield) [reply smarts/core/utils/file.py:118]: There is no guarantee that the pickle bytecode will be the same, only that it will result in the same object after unpickling. See here: https://death.andgravity.com/stable-hashing#pickle-not-stable

File: scenarios/sumo/figure_eight/scenario.py
Comment (saulfield) [reply scenarios/sumo/figure_eight/scenario.py:9]: Yes, since GLB generation now happens in `gen_scenario` (for caching purposes).

File: scenarios/sumo/intersections/2lane_circle/scenario.py
Comment (saulfield) [reply scenarios/sumo/intersections/2lane_circle/scenario.py:3]: Yes, I meant to refactor this to use `gen_scenario` instead. Thanks.

File: scenarios/sumo/zoo_intersection/scenario.py
Comment (saulfield) [reply scenarios/sumo/zoo_intersection/scenario.py]: Forgot to update this one too, thanks.

File: smarts/sstudio/__init__.py
Comment (saulfield) [reply smarts/sstudio/__init__.py:34]: If I'm understanding correctly, both of these options would make the internal `_gen_*` functions appears public, which I don't think should be the case.

File: docs/sim/scenario_studio.rst
Comment (saulfield) [reply docs/sim/scenario_studio.rst:50]: It seemed a bit redundant to add a `gen_scenario` example for each DSL object. What do you think?

File: smarts/sstudio/__init__.py
Comment (Gamenot) [reply smarts/sstudio/__init__.py:34]: I was considering a potential need for those methods, I am slightly hesitant to make them completely private but I do support removing them from `sstudio`. But I am unsure if we should still keep the public methods in `genscenario` for test-ability purposes.

File: smarts/sstudio/genscenario.py
Comment (Gamenot) [inline smarts/sstudio/genscenario.py:267]: Does the `overwrite` parameter make sense anymore for the `_gen` methods? It might just be reasonable to assume `True` always and remove the parameter.

File: smarts/sstudio/genscenario.py
Comment (Gamenot) [inline smarts/sstudio/genscenario.py:186]: Is this not just an error?

**Review by Gamenot — COMMENTED** [review]
I just have a few questions.

File: smarts/sstudio/__init__.py
Comment (saulfield) [reply smarts/sstudio/__init__.py:34]: Perhaps a middle ground would be to leave them out of `sstudio`, but make them public again? We now also print deprecation warnings when they are called from another module so the intended use is quite clear.

File: smarts/sstudio/genscenario.py
Comment (saulfield) [reply smarts/sstudio/genscenario.py:186]: This is how it worked before, I just kept the same behaviour. I'm open either way though. Is there a situation where there would be traffic histories and shifting the map would still be a valid thing to do?

File: smarts/sstudio/genscenario.py
Comment (saulfield) [reply smarts/sstudio/genscenario.py:117]: Yes, the overhead here is negligible. On the order of a few milliseconds in the worst case from what I remember.

File: smarts/sstudio/genscenario.py
Comment (saulfield) [reply smarts/sstudio/genscenario.py:106]: It would be, but this is for the parameters within the `Scenario` object. The comment is a bit ambiguous.

File: smarts/sstudio/genscenario.py
Comment (saulfield) [reply smarts/sstudio/genscenario.py:108]: You're right, I will pull it out into the conditionals where it is used.

File: smarts/sstudio/genscenario.py
Comment (saulfield) [reply smarts/sstudio/genscenario.py:267]: Yeah I agree, I'll make this change.

File: smarts/sstudio/genscenario.py
Comment (saulfield) [reply smarts/sstudio/genscenario.py:108]: On second thought, I think this actually is still useful, as it simplifies what the outer conditional would be. Without it, it would be like

```python
if (scenario.traffic and map_needs_rebuild) or _needs_build(
        db_conn, scenario.traffic, artifact_paths, obj_hash
    ):
    ...
```

rather than

```python
if _needs_build(
        db_conn, scenario.traffic, artifact_paths, obj_hash, map_needs_rebuild
    ):
    ...
```

I'll keep it like it was. Let me know if you disagree or if I'm missing a better way of doing this.

File: smarts/core/scenario.py
Comment (saulfield) [reply smarts/core/scenario.py:109]: I agree, these are all static methods though. What would you suggest in place of a property?

File: smarts/sstudio/genscenario.py
Comment (Gamenot) [reply smarts/sstudio/genscenario.py:186]: I do not think there is a case because it messes with the data. However, if we were to discard the underlying data this would still be OK. Maybe we leave as is.

File: smarts/sstudio/genscenario.py
Comment (Gamenot) [reply smarts/sstudio/genscenario.py:108]: Alright, that is a reasonable counter.

File: smarts/sstudio/genscenario.py
Comment (Gamenot) [reply smarts/sstudio/genscenario.py:117]: It could save a lot of time with thousands of scenarios. But I do not think there is a modern case for that yet.

**Review by Gamenot — APPROVED** [review]

File: smarts/sstudio/__init__.py
Comment (saulfield) [reply smarts/sstudio/__init__.py:34]: > Perhaps a middle ground would be to leave them out of `sstudio`, but make them public again? We now also print deprecation warnings when they are called from another module so the intended use is quite clear.

I made this change.

**Comment by saulfield** [comment]
Hi @qianyi-sun, getting to your comments just now.

> Hi Saul, just minor, have you tested the this in the replay scenarios (e.g., NGSIM)? I went though your changes/descriptions and the current scenario.py files in the NGSIM folder, looks ok but just wanted to double check.

They all build correctly using the new system. I haven't run any examples with them, but there shouldn't be any change.

> Also regarding the following point:
"The allow_offset_map option has been removed. This must now be set by setting shift_to_origin=True in a MapSpec object in the scenario.py file if this option is needed."
Should we check all the replay scenarios to see if we should add the shift_to_origin=True if we havent done yet? I can check them once this PR get merged.

Yes that's a good idea. There may be some scenarios that require that change that aren't covered by the tests. If you have some time to check them that would be helpful, thanks!

**Review by Adaickalavan — APPROVED** [review]

**Comment by Adaickalavan** [comment]
Should we put in a CI test to specifically test the scenario build caching feature?

**Comment by saulfield** [comment]
> Should we put in a CI test to specifically test the scenario build caching feature?

Good idea. I added a test in `cli/tests/test_studio.py`.

**Pull request merged**
